README.md
# bash-pseudo-hash

[![License](http://img.shields.io/badge/license-MIT-yellowgreen.svg)](#license)

Implements "fake" associative arrays suitable for use with bash versions that
precede 4.0 (with native hash table support) and zsh (as of v1.1.0).

## Build Status

| Service    | bash 4.3 | zsh 5.0.7 | zsh 5.3+ |
|:-----------|:--------:|:---------:|:--------:|
| Jenkins CI | [![Build Status](http://ci.mixtur.com/buildStatus/icon?job=bash-pseudo-hash+%28bash+4%29)]() | [![Build Status](http://ci.mixtur.com/buildStatus/icon?job=bash-pseudo-hash+%28zsh+5%29)]() | [![Build Status](http://ci.mixtur.com/buildStatus/icon?job=bash-pseudo-hash+%28zsh+5.3%2B%29)]() |

## Usage

Copy the script to a reasonable place in your project directory. Then include
the script in your own bash script by sourcing the file:

```sh
script_path="${BASH_SOURCE[0]}"; srcd="${srcd:-${(%):-%x}}"
base_path="$(builtin cd "$(dirname "${script_path}")" && builtin pwd)"
source "${base_path}/lib/bash_pseudo_hash.sh" || return 1
```

The above example places the `bash_pseudo_hash.sh` script in the `lib` directory
of your project.

### `RETVAL` return value convention

Values can be retrieved from __bash-pseudo-hash__ functions in two ways:

- variable assignment from the result of a sub-shell function call
- variable assignment from the `RETVAL` global variable following a function call
  (__recommended__)

All of the __bash-pseudo-hash__ functions will reset a global `RETVAL` variable
upon entry and update that variable before returning. This is similar to the
shell setting the `$?` return status global variable prior to returning from a
function call.

The advantage of leveraging the `RETVAL` global variable is that it avoids the
unnecessary forking of a subshell just so you can assign a return value to a
local variable. In other words: it's way more efficient.

Checking the value of the `RETVAL` global variable must occur immediately after
the associated function call returns as the value cannot be guaranteed to be
correct following another function call -- this behavior is exactly the same as
that for the `$?` return status global variable.

#### Create a hash

The following example creates a hash with two key/val pairs and uses sub-shell
function calls for variable assignment:

```sh
   local hash=()
   local key1="Phrase1"
   local val1="1000 pounds of spaghetti"
   local key2="Phrase2"
   local val2="And a bottle of beer."
   hash=( $(setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}") )
   hash=( $(setValueForKeyFakeAssocArray "${key2}" "${val2}" "${hash[*]}") )
```

The same example using `RETVAL` for variable assignment:

```sh
   local hash=()
   local key1="Phrase1"
   local val1="1000 pounds of spaghetti"
   local key2="Phrase2"
   local val2="And a bottle of beer."
   setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}" > /dev/null; hash=( ${RETVAL} )
   setValueForKeyFakeAssocArray "${key2}" "${val2}" "${hash[*]}" > /dev/null; hash=( ${RETVAL} )
```

#### Retrieve a value

Building on the previous example, here we retrieve our two values:

```sh
   local val1="$(valueForKeyFakeAssocArray "${key1}" "${hash[*]}")"
   local val2="$(valueForKeyFakeAssocArray "${key2}" "${hash[*]}")"
```

The same example using `RETVAL` for variable assignment:

```sh
   local val1="" val2=""
   valueForKeyFakeAssocArray "${key1}" "${hash[*]}" > /dev/null; val1="${RETVAL}"
   valueForKeyFakeAssocArray "${key1}" "${hash[*]}" > /dev/null; val2="${RETVAL}"
```

## Testing

Tests have been implemented for use with the [bash_unit](https://github.com/pgrange/bash_unit)
framework. To run the tests:

```sh
    prompt> bash_unit tests/*.sh
```

To run a specific test:

```sh
    prompt> bash_unit tests/test_set_keyval.sh
```

## Testing - zsh

Tests for zsh have been implemented for use with the [ZUnit](https://github.com/zunit-zsh/zunit)
framework. To run the tests:

```sh
    prompt> zunit tests/zunit/*.zunit
```

>NOTE: At this time, tests will fail on zsh 5.0 due to variable quoting bugs in
the ZUnit framework itself. Automated tests are currently using the following
forked branch: [markeissler/zunit](https://github.com/markeissler/zunit/tree/moe-fix-unquoted-vars).
A more detailed code review of the ZUnit source needs to be performed.

## Bugs and such

Submit bugs by opening an issue on this project's github page.

## Authors

__bash-pseudo-hash__ is the work of __Mark Eissler__.

## License

__bash-pseudo-hash__ is licensed under the MIT open source license.

---
Without open source, there would be no Internet as we know it today.


Changelog.md

v1.4.1 / 2017-10-29
===================

  * BPH-13 Tests fail on zsh 5.0.7 (var assignment bugs in ZUnit).
  * BPH-12 Fix encode/decode for zsh 5.3+.
  * BPH-11 Update README to revise Build Status to include bash/zsh versions.

v1.4.0 / 2017-10-22
===================

  * BPH-10 Update README to add zsh testing info.
  * BPH-9 Add zsh compatibility.

v1.3.0 / 2017-09-11
===================

  * Update README to remove hexdump dependency info.
  * Refactor __bphp_encode() to remove dependency on hexdump and improve
    performance.

v1.2.2 / 2017-09-10
===================

  * Fix valueForKeyFakeAssocArray() to return status 1 (value empty) or 2 (key
    not found).

v1.2.1 / 2017-09-10
===================

  * Fix valueForKeyFakeAssocArray() to return status 1 when key not found.

v1.2.0 / 2017-09-10
===================

  * Update README with RETVAL section.
  * Refactor to support RETVAL global var return value convention.

v1.1.1 / 2017-05-24
===================

  * Update README with Dependencies section.
  * Add __bph_version() function.

v1.1.0 / 2016-10-21
===================

  * Add prettyDumpFakeAssocArray() function.
  * Add keysForFakeAssocArray() function.
  * Add support for zsh.

v1.0.0 / 2016-09-20
===================

  * Add Changelog.


v0.9.0 / 2016-09-20
===================

  * First completed implementation.
  * Initial commit


tests/test_get_keys.sh
#!/usr/bin/env bash
g_path_script="$(builtin cd "$(dirname "${BASH_SOURCE[0]}")" && /bin/pwd)"
. "${g_path_script}/../bash_pseudo_hash.sh" || return 1

test_should_gets_keys() {
  local hash=()
  local testKeys="Phrase1 Phrase2 Phrase3"
  local testKeys_ary=()
  local expectedKeys="${testKeys}"
  local defaultIFS="$IFS"
  local IFS="$defaultIFS"

  IFS=$' ' testKeys_ary=( ${testKeys} ) IFS="$defaultIFS"

  for key in "${testKeys_ary[@]}"
  do
    local __val="100${key:(-1)} pounds of spaghettiü"
    hash=( $(setValueForKeyFakeAssocArray "${key}" "${__val}" "${hash[*]}") )
  done

  local returnedKeys="$(keysForFakeAssocArray "${hash[*]}")"
  assert_equals "${expectedKeys}" "${returnedKeys}"
}

test_should_gets_keys_RETVAL() {
  local hash=()
  local testKeys="Phrase1 Phrase2 Phrase3"
  local testKeys_ary=()
  local expectedKeys="${testKeys}"
  local defaultIFS="$IFS"
  local IFS="$defaultIFS"

  IFS=$' ' testKeys_ary=( ${testKeys} ) IFS="$defaultIFS"

  for key in "${testKeys_ary[@]}"
  do
    local __val="100${key:(-1)} pounds of spaghettiü"
    setValueForKeyFakeAssocArray "${key}" "${__val}" "${hash[*]}" > /dev/null
    hash="${RETVAL}"
  done

  local returnedKeys=""; keysForFakeAssocArray "${hash[*]}" > /dev/null; returnedKeys="${RETVAL}"
  assert_equals "${expectedKeys}" "${returnedKeys}"
}


tests/test_prettydump.sh
#!/usr/bin/env bash
g_path_script="$(builtin cd "$(dirname "${BASH_SOURCE[0]}")" && /bin/pwd)"
. "${g_path_script}/../bash_pseudo_hash.sh" || return 1

test_should_prettydump() {
  local hash=()
  local testKeys="Phrase1 Phrase2 Phrase3"
  local testKeys_ary=()
  local defaultIFS="$IFS"
  local IFS="$defaultIFS"

  IFS=$' ' testKeys_ary=( ${testKeys} ) IFS="$defaultIFS"

  for key in "${testKeys_ary[@]}"
  do
    local __val="100${key:(-1)} pounds of spaghettiü"
    hash=( $(setValueForKeyFakeAssocArray "${key}" "${__val}" "${hash[*]}") )
  done

  local _key _val
  local _i=1
  while read -r _key _val
  do
    local __expected="[Phrase${_i}]: 100${_i} pounds of spaghettiü"
    assert_equals "${__expected}" "${_key} ${_val}"
    (( _i++ ))
  done <<< "$(prettyDumpFakeAssocArray "${hash[*]}")"
}


tests/test_set_keyval.sh
#!/usr/bin/env bash
g_path_script="$(builtin cd "$(dirname "${BASH_SOURCE[0]}")" && /bin/pwd)"
. "${g_path_script}/../bash_pseudo_hash.sh" || return 1

test_should_set_keyval() {
    local hash=()
    local key1="Phrase1"
    local val1="1000 pounds of spaghettiü"
    hash=( $(setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}" ) )
    assert_equals "Phrase1:1000%20pounds%20of%20spaghetti%C3%BC" "${hash[0]}"
}

test_should_set_keyval_empty() {
    local hash=()
    local key1="Empty"
    local val1=""
    hash=( $(setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}" ) )
    assert_equals "Empty:%40%40" "${hash[0]}"
}

test_should_set_keyval_RETVAL() {
    local hash=()
    local key1="Phrase1"
    local val1="1000 pounds of spaghettiü"
    {
        setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}" > /dev/null
        hash=( ${RETVAL} )
    }
    assert_equals "Phrase1:1000%20pounds%20of%20spaghetti%C3%BC" "${hash[0]}"
}

test_should_set_keyval_empty_RETVAL() {
    local hash=()
    local key1="Empty"
    local val1=""
    {
        setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}" > /dev/null
        hash=( ${RETVAL} )
    }
    assert_equals "Empty:%40%40" "${hash[0]}"
}

test_should_get_keyval() {
    local hash=( "Phrase2:And%20a%20bottle%20of%20beer." )
    local key="Phrase2"
    local val="$(valueForKeyFakeAssocArray "${key}" "${hash[*]}")"
    assert_equals "And a bottle of beer." "${val}"
}

test_should_get_keyval_status() {
    local hash=( "Phrase2:And%20a%20bottle%20of%20beer." )
    local key="Phrase2"
    assert_status_code 0 "valueForKeyFakeAssocArray ${key} ${hash[*]}"
}

test_should_get_keyval_empty() {
    local hash=( "Empty:%40%40" )
    local key="Empty"
    local val="$(valueForKeyFakeAssocArray "${key}" "${hash[*]}")"
    assert_equals ""
}

test_should_get_keyval_empty_status() {
    local hash=( "Empty:%40%40" )
    local key="Empty"
    assert_status_code 1 "valueForKeyFakeAssocArray ${key} ${hash[*]}"
}

test_should_get_keyval_RETVAL() {
    local hash=( "Phrase2:And%20a%20bottle%20of%20beer." )
    local key="Phrase2"
    local val=""
    {
        valueForKeyFakeAssocArray "${key}" "${hash[*]}" > /dev/null
        val="${RETVAL}"
    }
    assert_equals "And a bottle of beer." "${val}"
}

test_should_get_keyval_empty_RETVAL() {
    local hash=( "Empty:%40%40" )
    local key="Empty"
    local val=""
    {
        valueForKeyFakeAssocArray "${key}" "${hash[*]}" > /dev/null
        val="${RETVAL}"
    }
    assert_equals ""
}

test_should_get_keyval_error() {
    local hash=( "Phrase2:And%20a%20bottle%20of%20beer." )
    local key="Phrase2x"
    assert_fail "valueForKeyFakeAssocArray ${key} ${hash[*]}"
}

test_should_get_keyval_error_status() {
    local hash=( "Phrase2:And%20a%20bottle%20of%20beer." )
    local key="Phrase2x"
    assert_status_code 2 "valueForKeyFakeAssocArray ${key} ${hash[*]}"
}


bash_pseudo_hash.sh
# bash_pseudo_hash.sh
#
# shellcheck shell=bash
# vi: set ft=bash
#
# Implements "fake" associative arrays suitable for use with bash(1) versions
# that precede 4.0 (with native hash table support) and zsh(1).
#
# Author : Mark Eissler (moe@markeissler.org) https://about.me/markeissler
# Website: https://github.com/markeissler/bash-pseudo-hash
#
# The MIT License (MIT)
#
# Copyright (C) 2015-2017 Mark Eissler. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

# source once and only once!
[[ ${BASH_PSEUDO_HASH:-} -eq 1 ]] && return || readonly BASH_PSEUDO_HASH=1

# zsh fixups
if [[ -n "${ZSH_VERSION// /}" ]]
then
    # add bash word split emulation for zsh
    #
    # see: http://zsh.sourceforge.net/FAQ/zshfaq03.html
    #
    setopt shwordsplit

    # force zsh to start arrays at index 0
    setopt KSH_ARRAYS
fi

# setValueForKeyFakeAssocArray()
# /*!
# @abstract Set value for key from a fake associative array
# @discussion
# Iterates over target_ary (an indexed array), searching for target_key, if the
#   key is found its value is set to new_value otherwise the target_key and
#   new_value are appended to the array.
#
#   The indexed array values must conform to this format:
#     "key:value"
#   Where key and value are separated by a single colon character.
#
#   Specify empty values as an empty, quoted string.
#
#   So-called "fake" associative arrays are useful for environments where the
#   installed version of bash(1) precedes 4.0.
# <pre>@textblock
#   hash=()
#   key1="Phrase1"
#   val1="1000 pounds of spaghetti"
#   key2="Phrase2"
#   val2="And a bottle of beer."
#   hash=( $(setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}") )
#   hash=( $(setValueForKeyFakeAssocArray "${key2}" "${val2}" "${hash[*]}") )
# @/textblock</pre>
#
#   Best practice is to use RETVAL for variable assignment:
# <pre>@textblock
#   hash=()
#   key1="Phrase1"
#   val1="1000 pounds of spaghetti"
#   key2="Phrase2"
#   val2="And a bottle of beer."
#   setValueForKeyFakeAssocArray "${key1}" "${val1}" "${hash[*]}" > /dev/null; hash=( ${RETVAL} )
#   setValueForKeyFakeAssocArray "${key2}" "${val2}" "${hash[*]}" > /dev/null; hash=( ${RETVAL} )
# @/textblock</pre>
# @param target_key Key to retrieve
# @param new_value New or updated value
# @param target_ary Indexed array to scan
# @return Returns new array with updated key (status 0) or an empty array
#   (status 1) on failure.
# @note Also sets global variable RETVAL to the same return value.
# */
setValueForKeyFakeAssocArray()
{
    # parameter list supports empty arguments!
    local target_key="${1}"; shift
    local new_value="${1//@/\\@}"; new_value="${new_value:-@@}"; shift # @todo: need to support setting nil values!
    local target_ary; target_ary=()
    local defaultIFS="$IFS"
    local IFS="$defaultIFS"
    local found=false
    unset RETVAL

    IFS=$' ' target_ary=( ${1} ) IFS="$defaultIFS"

    [[ -z "${target_key// /}" ]] && RETVAL="" && echo "${RETVAL}" && return 1

    local _target_ary_length=${#target_ary[@]}
    local _encoded_new_value=""
    __bphp_encode "${new_value}" > /dev/null; _encoded_new_value="${RETVAL}"
    local i
    for (( i=0; i<${_target_ary_length}; i++ ))
    do
        local __val="${target_ary[i]}"

        if [[ "${__val%%:*}" == "${target_key}" ]]
        then
            target_ary[$i]="${__val%%:*}:${_encoded_new_value}"
            found=true
            break
        fi

        unset __val
    done
    unset i _target_ary_length

    # key not found, append
    [[ "${found}" == false ]] && target_ary+=( "${target_key}:${_encoded_new_value}" )

    RETVAL="${target_ary[*]}"

    echo "${RETVAL}" && return 0
}

# valueForKeyFakeAssocArray()
# /*!
# @abstract Fetch value for key from a fake associative array
# @discussion
# Iterates over target_ary (an indexed array), searching for target_key, if the
#   key is found its value is returned.
#
#   The indexed array values must conform to this format:
#     "key:value"
#   Where key and value are separated by a single colon character.
#
#   So-called "fake" associative arrays are useful for environments where the
#   installed version of bash(1) precedes 4.0.
# <pre>@textblock
#   hash=() # hash returned previously by setValueForKeyFakeAssocArray()
#   key1="Phrase1"
#   key2="Phrase2"
#   val1="$(valueForKeyFakeAssocArray "${key1}" "${hash[*]}")"
#   val2="$(valueForKeyFakeAssocArray "${key2}" "${hash[*]}")"
# @/textblock</pre>
#
#   Best practice is to use RETVAL for variable assignment:
# <pre>@textblock
#   hash=() # hash returned previously by setValueForKeyFakeAssocArray()
#   key1="Phrase1"
#   key2="Phrase2"
#   local val1="" val2=""
#   valueForKeyFakeAssocArray "${key1}" "${hash[*]}" > /dev/null; val1="${RETVAL}"
#   valueForKeyFakeAssocArray "${key1}" "${hash[*]}" > /dev/null; val2="${RETVAL}"
# @/textblock</pre>
# @param target_key Key to retrieve
# @param target_ary Indexed array to scan
# @return Returns string containing value (status 0 if non-empty, status 1 if
#   empty) or an empty string (status 2) on key not found or target_key is empty.
# @note Also sets global variable RETVAL to the same return value.
# */
valueForKeyFakeAssocArray()
{
    local target_key="${1}"; shift
    local target_ary; target_ary=()
    local defaultIFS="$IFS"
    local IFS="$defaultIFS"
    local value=""
    local found_key=false
    unset RETVAL

    IFS=$' ' target_ary=( ${1} ) IFS="$defaultIFS"

    [[ -z "${target_key// /}" || ${#target_ary[@]} -eq 0 ]] && RETVAL="" && echo "${RETVAL}" && return 2

    local _item
    for _item in "${target_ary[@]}"
    do
        if [[ "${_item%%:*}" == "${target_key}" ]]
        then
            found_key=true
            # @todo: need to support returning nil values!
            __bphp_decode "${_item#*:}" > /dev/null; value="${RETVAL}"
            # decode nils
            value="${value//@@/}"; value="${value//\\@/@}"
            break
        fi
    done
    unset _item

    if [[ "${found_key}" == false ]]
    then
        RETVAL="" && echo "${RETVAL}"
        return 2
    elif [[ -z "${value// /}" ]]
    then
        RETVAL="" && echo "${RETVAL}"
        return 1
    fi

    echo "${RETVAL}" && return 0
}

# keysForFakeAssocArray()
# /*!
# @abstract Returns list of keys in fake associative array
# @discussion
# Iterates over target_ary (an indexed array) and extracts key component from
#   each element.
# @param target_ary Indexed array to scan
# @return Returns string of space separated keys (status 0) or an empty string
#   (status 1) on failure.
# @note Also sets global variable RETVAL to the same return value.
# */
keysForFakeAssocArray() {
    local target_ary; target_ary=()
    local target_ary_keys; target_ary_keys=()
    local defaultIFS="$IFS"
    local IFS="$defaultIFS"
    unset RETVAL

    IFS=$' ' target_ary=( ${1} ) IFS="$defaultIFS"

    [[ ${#target_ary[@]} -eq 0 ]] && RETVAL="" && echo "${RETVAL}" && return 1

    local _item
    for _item in "${target_ary[@]}"
    do
        _item="${_item}"
        target_ary_keys+=( "${_item%%:*}" )
    done
    unset _item

    if [[ "${#target_ary_keys[@]}" -eq 0 ]]
    then
        RETVAL="" && echo "${RETVAL}"
        return 1
    fi

    RETVAL="${target_ary_keys[*]}"

    echo "${RETVAL}" && return 0
}

# prettyDumpFakeAssocArray()
# /*!
# @abstract Output a pretty-printed dump of fake associative array contents
# @discussion
# Iterates over target_ary (an indexed array) and extracts keys and values from
#   each element. Pretty prints the output in the following format:
# <pre>@textblock
#   [key]: value
# /@textblock</pre>
# @param target_ary Indexed array to scan
# @return Returns formatted string of key/values (status 0) or an empty string
#   (status 1) on failure.
# */
prettyDumpFakeAssocArray() {
    local target_ary; target_ary=()
    local defaultIFS="$IFS"
    local IFS="$defaultIFS"

    IFS=$' ' target_ary=( ${1} ) IFS="$defaultIFS"

    [[ ${#target_ary[@]} -eq 0 ]] && echo "" && return 1

    local _item
    for _item in "${target_ary[@]}"
    do
        local __key="${_item%%:*}"
        local __val=""
        valueForKeyFakeAssocArray "${__key}" "${target_ary[*]}" > /dev/null; __val="${RETVAL}"
        printf "  [%s]: %s\n" "${__key}" "${__val}"
        unset __key __val
    done
    unset _item
}

# __bphp_encode()
# /*!
# @internal
# */
__bphp_encode()
{
    local string="${1}"
    local new_string=""
    local LANG=C
    unset RETVAL

    [[ -z "${string// /}" ]] && RETVAL="" && echo "${RETVAL}" && return 1

    local _string_len=${#string}
    local i
    for (( i = 0; i<${_string_len}; i++ ))
    do
        local __char="${string:$i:1}"
        case $__char in
            [A-Za-z0-9.~_-] )
                new_string+="$__char"
                ;;
            * )
                if [[ -n "${ZSH_VERSION// /}" ]]
                then
                    if [[ -z "${ZSH_VERSION_ARY[@]}" ]]
                    then
                        __bphp_parse_semver "${ZSH_VERSION}" > /dev/null
                        ZSH_VERSION_ARY=( ${RETVAL} )
                    fi
                    if [[ ${ZSH_VERSION_ARY[0]} -gt 5 ]] || [[ ${ZSH_VERSION_ARY[0]} -eq 5 && ${ZSH_VERSION_ARY[1]} -ge 3 ]]
                    then
                        printf -v __char '\\x%02X' "'$__char"
                    else
                        __char="$(printf '\\x%02X' "'$__char")"
                    fi
                else
                    printf -v __char '\\x%02X' "'$__char"
                fi
                new_string+="${__char//\\x/%}"
                ;;
        esac
    done
    unset i _string_len

    if [[ -z "${new_string// /}" ]]
    then
        RETVAL="" && echo "${RETVAL}"
        return 1
    fi

    RETVAL="${new_string}"

    echo "${RETVAL}" && return 0
}

# __bphp_decode()
# /*!
# @internal
# */
__bphp_decode()
{
    local string="${1}"
    local new_string
    unset RETVAL

    [[ -z "${string// /}" ]] && RETVAL="" && echo "${RETVAL}" && return 1

    if [[ -n "${ZSH_VERSION// /}" ]]
    then
        if [[ -z "${ZSH_VERSION_ARY[@]}" ]]
        then
            __bphp_parse_semver "${ZSH_VERSION}" > /dev/null
            ZSH_VERSION_ARY=( ${RETVAL} )
        fi
        if [[ ${ZSH_VERSION_ARY[0]} -gt 5 ]] || [[ ${ZSH_VERSION_ARY[0]} -eq 5 && ${ZSH_VERSION_ARY[1]} -ge 3 ]]
        then
            printf -v new_string "%b" "${string//\%/\\x}"
        else
            new_string="$(printf "%b" "${string//\%/\\x}")"
        fi
    else
        printf -v new_string "%b" "${string//\%/\\x}"
    fi

    if [[ -z "${new_string// /}" ]]
    then
        RETVAL="" && echo "${RETVAL}"
        return 1
    fi

    RETVAL="${new_string}"

    echo "${RETVAL}" && return 0
}

__bphp_parse_semver()
{
    local string="${1}"
    local semver_ary; semver_ary=()
    unset RETVAL

    [[ -z "${string// /}" ]] && RETVAL="" && echo "${RETVAL}" && return 1

    local major="${string%%.*}"; major="${major:-0}"
    local minor="${string%.*}"; minor="${minor#*.}"; minor="${minor:-0}"
    local patch="${string##*.}"; patch="${patch:-0}"

    semver_ary=( "${major}" "${minor}" "${patch}" )

    if [[ ${#semver_ary[@]} -eq 0 ]]
    then
        RETVAL=""; echo "${RETVAL}"; return 1
    fi

    RETVAL="${semver_ary[*]}"; echo "${RETVAL}"; return 0
}

__bph_version()
{
    local version="1.4.0"

    echo "${version}" && return 0
}
